Add util::Expected (std::expected)

maflcko opened the pull request:
Some low-level code could benefit from being able to use `std::expected` from C++23:

* Currently, some code is using `std::optional<E>` to denote an optional error. This is fine, but a bit confusing, because `std::optional` is normally used for values, not errors. Using `std::expected<void, E>` is clearer.
* Currently, some code is using `std::variant<V, E>` to denote either a value or an error. This is fine, but a bit verbose, because `std::variant` requires a visitor or get_if/holds_alternative instead of a simple call of the `operator bool` for `std::expected`.

In theory, `util::Result` could be taught to behave similar to `std::expected` (see https://github.com/bitcoin/bitcoin/pull/34005). However, it is unclear if this is the right approach:

* `util::Result` is mostly meant for higher level code, where errors come with translated error messages.
* `std::expected` is mostly meant for lower level code, where errors could be an enum, or any other type.
* https://github.com/bitcoin/bitcoin/pull/25665 aims to minimize the memory footprint of the error by wrapping it in a unique_ptr internally. `std::expected` requires the value and error to be "nested within it" (https://cplusplus.github.io/LWG/issue4141). So from a memory-layout perspective, the two are not compatible.
* `std::expected` also comes with `std::unexpected`, which also does not map cleanly to `util::Result`.

So just add a minimal drop-in port of `std::expected`.

Commit:
Add util::Expected (std::expected)

purpleKarrot commented:
> Some low-level code could benefit from being able to use `std::expected` from C++23.

Can you elaborate? How would it benefit? Over what alternative?

stickies-v reviewed (commented):
Concept ACK. Functions with return type `std::optional<E>` returning a truthy value on failure is unintuitive, this is much better.

hodlinator reviewed (commented):
Concept ACK fa1eee4f46fd089d6e19b124addee7c7679585f4

I prefer introducing `util::Expected` as a distinct type from `util::Result`, with a clear upgrade path to `std::expected` once we have C++23. (Maybe `util::Result` can later be implemented as a special case of `util::Expected`).

Recently suggested `std::variant` in another PR instead of using out-parameters (https://github.com/bitcoin/bitcoin/pull/33657#discussion_r2584393520) but would strongly prefer `util::Expected`, had it existed.

ryanofsky commented on src/util/expected.h:
```diff
@@ -0,0 +1,78 @@
+class Expected
+{
+private:
+    using T = std::conditional_t<std::is_same_v<V, void>, std::monostate, V>;
```
In commit "Add util::Expected (std::expected)" (fa1eee4f46fd089d6e19b124addee7c7679585f4)

Normally T is a template parameter and internal types like this have more descriptive names. Would suggest renaming V to T to match std::expected [documentation](https://en.cppreference.com/w/cpp/utility/expected.html) and T to ValueType or similar

maflcko replied:
thx, done

ryanofsky commented on src/util/result.h:
```diff
@@ -21,9 +21,9 @@ struct Error {
-//! classes like std::optional, std::variant, and std::tuple, or custom structs
-//! and enum types to return function results.
+//! that only need error-handling should avoid util::Result and instead use
+//! util::Expected, or custom structs and enum types to return function
```
In commit "Add util::Expected (std::expected)" (fa1eee4f46fd089d6e19b124addee7c7679585f4)

IMO std::optional and std::variant are still appropriate to use in many cases and it would nice to keep mentioning them.

ryanofsky reviewed (approved):
Code review ACK fa80cca5dd193668300df891876287dacca6159e. Looks great! This is a minimal implementation of the std::expected class that may be marginally less efficient (doesn't have a void specialization, may do some unnecessary moves/copies) and lacks some features, but is usable and can be expanded, and is basically as simple as possible. I left some comments and suggestions below but none are important. I also updated the #25665 PR description to compare the `Result` and `Expected` classes.

maflcko commented on src/net_processing.h:
```diff
@@ -103,7 +104,7 @@ class PeerManager : public CValidationInterface, public NetEventsInterface
      * @param[in]  block_index  The blockindex
      * @returns std::nullopt if a request was successfully made, otherwise an error message
```
thx, removed this line

ryanofsky reviewed (approved):
Code review ACK fa76a6620012fb738639d8fd7ce17b185bfd376c

stickies-v reviewed (commented):
Approach ACK fa76a6620012fb738639d8fd7ce17b185bfd376c

arejula27 reviewed (commented):
Concept ACK [fa76a6620012fb738639d8fd7ce17b185bfd376c]

This PR is valuable because it creates a smooth path toward adopting the new C++23 standard. Instead of requiring a full refactor to `std::expected` all at once, it allows the codebase to transition gradually, simply by replacing `utils::expected` with `std::expected` later on.

I also like the idea of introducing this error handling paradigm into the repository. `expected` is a clean way to handle errors and aligns with the modern programming languages such as Rust, where a function returns either a value or an error. By contrast, `try`/`catch` approach (also mentioned in the discussion)  often lead to poor error handling, either vague error messages or overly complex and hard-to-reason-about control flow. Using an explicit result type avoids these issues.

`std::optional` is  currently used in this repo  as an approximation of this pattern, but it isn’t a correct fit. A function can succeed with a value, succeed without a value, or fail with an error, all of which `optional` cannot distinguish cleanly.

Looking forward, I’d like to consider using only `std::expected` throughout the codebase, with `utils::Result` becoming a thin wrapper around `std::expected` only when additional logging is required. Alternatively, instead of `utils::Result` we could provide static helper functions, which consumes `std::expected` to make error logging  user-friendly. Unifying the error-handling approach would make maintenance and review easier, since developers wouldn’t need to think about which mechanism is appropriate for each function, there would be just one.

However, as I mentioned in an inline comment, there are also good arguments in favor of keeping both types. If we go in that direction, I would suggest improving the documentation and adding a constructor (or conversion path) from `utils::expected` to `utils::Result`. This would make it easier to use both patterns together: `utils::expected` for nested or lower-level functions, and `utils::Result` only at the top level where a user-facing error message is ultimately needed.

maflcko commented on src/net_processing.h:
```diff
@@ -11,6 +11,7 @@
 #include <node/txorphanage.h>
 #include <protocol.h>
 #include <threadsafety.h>
+#include <util/expected.h>
```
> I wouldn’t modify `src/net_processing.h` in this PR. Even though it’s a good showcase of the improvement, there are other functions that still use `std::optional` for errors, for example, line 828 in `src/netbase.cpp`.

No, it is not. This is a valid and intended use of `std::optional`:

```
    std::optional<CNetAddr> addr{LookupHost(str_addr, /*fAllowLookup=*/false)};
    if (addr.has_value()) {
```

I'll leave as-is for now. Also, I'll keep touching the file, so that a valid and intended use of `std::expected` can be seen.

Sjors replied:
`FetchBlock()` is only used by the `getblockfrompeer` RPC (which I introduced in #20295), so it's a nice and fairly isolated example to use.

Typically we gradually replace old code with more modern patterns while editing something in the same area, so the other examples will follow later.

maflcko commented on src/util/result.h:
```diff
@@ -21,8 +21,8 @@ struct Error {
 //!
 //! It is intended for high-level functions that need to report error strings to
```
> However, Result in our codebase isn’t limited to reporting errors to end users. As @ryanofsky noted in PR #25665, there are broader use cases. One strong example is:
> 
> > Result may also be a better choice than std::expected when returning pointer values. (Followup #26022 adds a ResultPtr specialization that avoids double dereferencing.)

Thx, this is a good point. However, I think modifying the Result and Expected docs for ResultPtr is unrelated to this pull and can be done in the pull introducing ResultPtr.

I'll leave as-is for now.

maflcko replied:
> I would like you to consider adding **directly** as I mentioned, as you already did some small text corrections on `results.h`

Not sure. I am not touching this line, so I'll leave as-is for now.

maflcko commented on src/util/expected.h:
```diff
@@ -0,0 +1,81 @@
+{
+public:
+    constexpr explicit Unexpected(E e) : err(std::move(e)) {}
+    E err;
```
I know, and this was intentional. I can't imagine a real code snippet in this code-base that uses `.err` or `.error()`. The only use-case of Unexpected is to serve as a marker, which Expected constructor to pick internally. Am I missing something?

maflcko replied:
thx, pushed in https://github.com/bitcoin/bitcoin/pull/34032/commits, fa7ceec5b07fbc99c4cc247803f78fb8b019c138

sedited reviewed (approved):
ACK fa76a6620012fb738639d8fd7ce17b185bfd376c

maflcko commented on src/util/expected.h:
```diff
@@ -0,0 +1,81 @@
+
+    constexpr const ValueType& value() const LIFETIMEBOUND
+    {
+        assert(has_value());
```
I see your point, but exactly the same scenario can happen without this pull, when an `util::Result` is replaced by `std::expected`. (And the inverse can happen when going the other way)

I don't feel strongly, but I think it is minimally safer to assert here, because for our code-base it is a programming logic error to call `value` when a value does not exist. So it seems minimally better to be able to detect those cases more easily while we have Expected (before switching to `std::expected`)


I understand the stdlib went a different path, looking at https://en.cppreference.com/w/cpp/experimental/optional/bad_optional_access.html, which initially inherited from logic_error, but later (https://en.cppreference.com/w/cpp/utility/bad_optional_access.html) it was only derived from std::exception.


In any case, I don't feel strongly, and I don't think (and hope) any of this will matter in practise, and I am happy to push this patch, if you insist:

```diff
diff --git a/src/util/expected.h b/src/util/expected.h
index 6938c97..081c263 100644
--- a/src/util/expected.h
+++ b/src/util/expected.h
@@ -8,6 +8,7 @@
 #include <attributes.h>
 
 #include <cassert>
+#include <exception>
 #include <type_traits>
 #include <utility>
 #include <variant>
@@ -24,6 +25,11 @@ public:
     E err;
 };
 
+struct BadExpectedAccess : std::exception
+{
+    const char* what() const noexcept override { return "Bad util::Expected access"; }
+};
+
 /// The util::Expected class provides a standard way for low-level functions to
 /// return either error values or result values.
 ///
@@ -49,12 +55,16 @@ public:
 
     constexpr const ValueType& value() const LIFETIMEBOUND
     {
-        assert(has_value());
+        if (!has_value()) {
+            throw BadExpectedAccess{};
+        }
         return std::get<0>(m_data);
     }
     constexpr ValueType& value() LIFETIMEBOUND
     {
-        assert(has_value());
+        if (!has_value()) {
+            throw BadExpectedAccess{};
+        }
         return std::get<0>(m_data);
     }

maflcko commented on src/net_processing.h:
```diff
@@ -101,9 +102,8 @@ class PeerManager : public CValidationInterface, public NetEventsInterface
-     * @returns std::nullopt if a request was successfully made, otherwise an error message
      */
-    virtual std::optional<std::string> FetchBlock(NodeId peer_id, const CBlockIndex& block_index) = 0;
+    virtual util::Expected<void, std::string> FetchBlock(NodeId peer_id, const CBlockIndex& block_index) = 0;
```
Thanks. Very nice, pushed a commit to do that.

Commit:
refactor: Enable clang-tidy bugprone-unused-return-value

This requires some small refactors to silence false-positive warnings.

Also, expand the bugprone-unused-return-value.CheckedReturnTypes option
to include util::Result, and util::Expected.

romanz reviewed (commented):
tACK faa23738fc

Works with #33657 (rebased over this PR).

sedited reviewed (approved):
Re-ACK faa23738fc2576e412edb04a4004fab537a3098e

hodlinator reviewed (approved):
ACK faa23738fc2576e412edb04a4004fab537a3098e

rkrux reviewed (commented):
light Code Review ACK faa23738fc2576e412edb04a4004fab537a3098e

Onboard with the intent of using (util/std)::Expected for the use cases mentioned in the PR description.

ryanofsky reviewed (approved):
Code review ACK faa23738fc2576e412edb04a4004fab537a3098e, only adding `value_or` implementation and `CheckedReturnTypes` clang-tidy commit since last review.

I think this is ok to merge as-is but I left some suggestions about `value_or`, and I think the `[[maybe_unused]] auto _{...}` uses in the second commit are pretty ugly and it would be nice if they could be changed to use `(void)` instead.

It could also be nice make the other suggested improvements, but not important:

- Avoiding returning `std::monostate` as suggested by hodlinator: https://github.com/bitcoin/bitcoin/pull/34006#discussion_r2592454740
- Throwing BadExpectedAccess in release builds, aborting in debug builds as suggested by hodlinator https://github.com/bitcoin/bitcoin/pull/34006#discussion_r2597679959 seems reasonable
- Adding Unexpected accessor as suggested by stickies https://github.com/bitcoin/bitcoin/pull/34006#discussion_r2593341384 also seems reasonable
- Using in_place_index suggested by stickies https://github.com/bitcoin/bitcoin/pull/34006#discussion_r2598773353
- Improving httpserver as suggested by stickies https://github.com/bitcoin/bitcoin/pull/34006#discussion_r2599122933

EDIT: Linked to more suggestions

stickies-v reviewed (approved):
ACK faa23738fc2576e412edb04a4004fab537a3098e

Current code seems fine to merge as-is, but I would prefer addressing more of the outstanding comments (from myself and other reviewers) first.

ryanofsky commented:
I merged this since it seems like a clear improvement that other changes can build on. Also took "whichever happens earlier" (https://github.com/bitcoin/bitcoin/pull/34006#discussion_r2599410546) as a hint to mean the author is happy to see this merged as is. Followup suggestion are listed in: https://github.com/bitcoin/bitcoin/pull/34006#pullrequestreview-3552578391

maflcko commented on src/httpserver.cpp:
```diff
@@ -330,7 +330,7 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
         assert(g_work_queue);
         if (g_work_queue->Enqueue(item.get())) {
-            item.release(); /* if true, queue took ownership */
+            [[maybe_unused]] auto _{item.release()}; /* if true, queue took ownership */
```
left as-is for now

maflcko commented:
I pushed https://github.com/bitcoin/bitcoin/pull/34032 to add all of the unused member functions requested by reviewers, and resolved the threads here. let me know if I have forgotten anything.
